sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/decommission/DecommissionPlanFactory.java

```diff
@@ -0,0 +1,289 @@
+ * <li>Kill is issued for all tasks in the pod.</li>
+ * <li>As resources for tasks within the pod are offered, the DECOMMISSIONED+IN_PROGRESS state results in those
+ * resources being unreserved and removed from the {@code TaskInfo}.</li>
+ * <li>Once the resources are all cleared, the task is deleted from the {@link StateStore}.</li></ol>
+ *
+ * Note that this is different from uninstall behavior in a couple ways, resulting in different handling from uninstall:
+ *
+ * <ul><li>The decommission operation can be cancelled by incrementing pod count, whereas uninstall cannot be cancelled
+ * once it's started. We remove resources from the {@code TaskInfo}s instead of marking them with tombstones to reduce
+ * the likelihood of a partial decommission causing problems.</li>
+ * <li>The decommission operation occurs serially in a predictable order, whereas uninstall attempts to unreserve all
+ * known resources in parallel. We enforce this by having the decommission plan set the DECOMMISSIONED+IN_PROGRESS bit
+ * one pod at a time, and only unreserve pods that are marked as DECOMMISSIONED+IN_PROGRESS.</li></ul>
+ */
+public class DecommissionPlanFactory {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DecommissionPlanFactory.class);
+
+    /**
+     * The status for a task whose resources should be unreserved because its resources are currently being
+     * decommissioned.
+     */
+    public static final GoalStateOverride.Status DECOMMISSIONING_STATUS =
+            GoalStateOverride.DECOMMISSIONED.newStatus(GoalStateOverride.Progress.IN_PROGRESS);
+
+    private final PlanInfo planInfo;
+
+    public DecommissionPlanFactory(ServiceSpec serviceSpec, StateStore stateStore, TaskKiller taskKiller) {
+        this.planInfo = buildPlanInfo(serviceSpec, stateStore, taskKiller);
+    }
+
+    /**
+     * Returns a {@link Plan} for decommissioning tasks, or an empty {@link Optional} if nothing needs to be
+     * decommissioned.
+     */
+    public Optional<Plan> getPlan() {
+        return planInfo.plan;
+    }
+
+    /**
+     * Returns all {@link ResourceCleanupStep}s associated with the decommission plan, or an empty list if no steps are
+     * applicable.
+     */
+    public Collection<Step> getResourceSteps() {
+        return planInfo.resourceSteps;
+    }
+
+    private static class PlanInfo {
+        private final Optional<Plan> plan;
+        private final Collection<Step> resourceSteps;
+
+        private PlanInfo(Optional<Plan> plan, Collection<Step> resourceSteps) {
+            this.plan = plan;
+            this.resourceSteps = resourceSteps;
+        }
+    }
+
+    /**
+     * Returns a {@link Plan} for decommissioning tasks, or an empty {@link Optional} if no decommission is necessary.
+     */
+    private static PlanInfo buildPlanInfo(ServiceSpec serviceSpec, StateStore stateStore, TaskKiller taskKiller) {
+        // Determine which tasks should be decommissioned (and which shouldn't)
```

Comment:
Note to self: Decommissioning is done at the pod level. So, a single task being decommissioned should result in the pod going down.